### PR TITLE
[Bug Fix] Enum Serialization

### LINKED ISSUE #5983: Enums functionality fails when memory limit is low
### What happens?

Converting a table with varchar columns to a table with enum columns increases the size of a duckdb database by more than double. The operation does not complete either. This happens when the size of the database is close to the memory limit of the duckdb persistent session.
### To Reproduce

```bash
./build/debug/duckdb mem_limit_enum.duckdb
```

```SQL
set memory_limit='2GB';
load tpch;
call dbgen(sf=2);
SELECT count(distinct(l_comment)) from lineitem;
CREATE TYPE l_comment_enum as ENUM(select l_comment from lineitem); # statement 1
CREATE TABLE lineitem2 (comment l_comment_enum); # statement 2
```
After statement 2, the creation of the table increases the size of the database by over 50%.
You can keep creating tables to show this
```
INSERT INTO lineitem2 (select l_comment from lineitem);
```
Observe as the size of mem_limit_enum.duckdb increases to over 5GB and the insert statement does not complete.

### OS:

MacOs Monterey

### DuckDB Version:

v0.6.2-dev1218

### DuckDB Client:

CLI

### Full Name:

Tom Ebergen

### Affiliation:

DuckDB labs

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### PR DESCRIPTION
**Pull request by pdet**
This PR solves the issues related to Enum types being serialized on the catalog, at every row group, and every usage in a table.

We now only serialize the Enum at the catalog and properly reconstruct the pointers to the catalog during loading. The only exception is if the Enum comes from an outside object scan (e.g., Parquet, Arrow, Pandas, R, ...), then they are serialized in the table since there is no catalog entry.

In addition, I also rewrote the code that creates ENUMs from query results; the old code was unnecessarily reading the data 3x.

Fixes: #5983

### CODE REVIEW

**Comment by Mytherin** [comment]
Retargeting this to feature as it breaks the storage version

**Comment by pdet** [comment]
@Mytherin maybe we need some clicks here to restart CI?

**Comment by Mytherin** [comment]
I can't restart CI if it has never started - perhaps merge with feature?

File: src/common/types.cpp [hunk 1]
```diff
@@ -1523,20 +1578,68 @@ shared_ptr<ExtraTypeInfo> ExtraTypeInfo::Deserialize(FieldReader &reader) {
...
+			if (catalog) {
+				auto enum_type = catalog->GetType(client_context, schema_name, enum_name, true);
+				if (enum_type != LogicalType::INVALID) {
+					extra_info = enum_type.GetAuxInfoShrPtr();
+					break;
+				}
+			}
+			// Try the DB Path as a catalog name
+			auto catalog_name = AttachedDatabase::ExtractDatabaseName(client_context.db->config.options.database_path);
+			if (Catalog::TypeExists(client_context, catalog_name, schema_name, enum_name)) {
```
Comment (Mytherin) [inline src/common/types.cpp:1598]: The logic here seems to duplicate the standard catalog resolution logic - can we just make this `INVALID_CATALOG` instead and let the catalog handle where to look up the enum?

Also - is all of this necessary in the first place? Shouldn't there just always be a catalog that we are deserializing from?

File: src/common/types.cpp [hunk 2]
```diff
@@ -1336,12 +1350,40 @@ struct EnumTypeInfo : public ExtraTypeInfo {
...
+	// Store If we are serializing the internals
+	writer.WriteField<bool>(serialize_internals);
+	if (serialize_internals) {
+		// We must serialize the internals
+		auto dict_size = enum_info.GetDictSize();
+		// Store Dictionary Size
+		writer.WriteField<uint32_t>(dict_size);
+		// Store Vector Order By Insertion
+		((Vector &)enum_info.GetValuesInsertOrder()).Serialize(dict_size, writer.GetSerializer());
+		;
```
Comment (Mytherin) [inline src/common/types.cpp:1383]: nit: empty statement

**Review by Mytherin — COMMENTED** [review]
Thanks for the fixes - looks good. Some more minor comments:

File: src/common/types.cpp (hunk 1 above)
Comment (pdet) [reply src/common/types.cpp:1598]: Not always. At the ` Planner::VerifyPlan`, there is no catalog, for example.
Indeed using `INVALID_CATALOG` does the trick there :-)

**Review by Mytherin — COMMENTED** [review]
Thanks! Looks good - could you just merge with master and regenerate the storage version?

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**